client/network/src/bitswap.rs

```diff
@@ -0,0 +1,337 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Bitswap server for substrate.
+//!
+//! Allows querying transactions by hash over standard bitswap protocol
+//! Only supports bitswap 1.2.0.
+//! CID is expected to reference 256-bit Blake2b transaction hash.
+
+use std::collections::VecDeque;
+use std::io;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use cid::Version;
+use codec::Encode;
+use core::pin::Pin;
+use futures::Future;
+use futures::io::{AsyncRead, AsyncWrite};
+use libp2p::core::{
+	connection::ConnectionId, Multiaddr, PeerId,
+	upgrade, InboundUpgrade, OutboundUpgrade, UpgradeInfo,
+};
+use libp2p::swarm::{
+	NetworkBehaviour, NetworkBehaviourAction, NotifyHandler, PollParameters,
+	ProtocolsHandler, IntoProtocolsHandler, OneShotHandler,
+};
+use log::{error, debug, trace};
+use prost::Message;
+use sp_runtime::traits::{Block as BlockT};
+use unsigned_varint::{encode as varint_encode};
+use crate::chain::Client;
+use crate::schema::bitswap::{
+	Message as BitswapMessage,
+	message::{wantlist::WantType, Block as MessageBlock, BlockPresenceType, BlockPresence},
+};
+
+const LOG_TARGET: &str = "bitswap";
+
+// Undocumented, but according to JS the bitswap messages have a max size of 512*1024 bytes
+// https://github.com/ipfs/js-ipfs-bitswap/blob/
+// d8f80408aadab94c962f6b88f343eb9f39fa0fcc/src/decision-engine/index.js#L16
+// We set it to the same value as max substrate protocol message
+const MAX_PACKET_SIZE: usize = 16 * 1024 * 1024;
```

Comment:
This is not a protocol limit. Just some arbitrary value chosen by the JS implementation. Furthermore it seems like this is outbound limit and I'm not sure what' the receiving limit is. If someone needs to be able to retrieve huge transactions they'll just have to use an implementation that does not have that limit for now.

> I wonder a bit if big tx should be split in multiple block.

Each block may only be retrieved by its hash. In case of transactions that equals transaction hash. Splitting would require some kind of container chema that would complicate things a lot.

Reply:
That's fine then. (good to hear splitting support may not be needed (I agree it is not a trivial change :))